jwt/jwk.py

```diff
@@ -279,25 +276,139 @@ def jwk_from_dict(dct: Mapping[str, Any]) -> AbstractJWKBase:
     return supported[kty].from_dict(dct)
 
 
-def jwk_from_pem(pem_content: bytes) -> AbstractJWKBase:
-    if not isinstance(pem_content, bytes):  # pragma: no cover
-        raise TypeError('pem_content must be a bytes')
-
+def jwk_from_bytes_argument_conversion(func):
+    if 'private' in func.__name__:
+        kind = 'private'
+    elif 'public' in func.__name__:
+        kind = 'public'
+    else:
+        raise Exception("the wrapped function must have either public"
+                        " or private in it's name")
+
+    def wrapper(content, **kwargs):
+        # content is the only positional argument allowed
+        # the function you wrap must use kwargs-only arguments PEP3102
+        if not isinstance(content, bytes):  # pragma: no cover
+            raise TypeError(
+                'content must be bytes, it is {}'.format(type(content)))
+        load_function = f'serializer_load_function_{kind}'
+        if kwargs.get(load_function) is None:
+            raise Exception(f'{load_function} is not Optional')
+        # now convert it to a Callable if it's a string
+        if isinstance(kwargs.get(load_function), str):
+            kwargs[load_function] = getattr(
+                serialization_module,
+                kwargs.get(load_function),
+            )
+
+        if kwargs.get('options') is None:
+            kwargs['options'] = {}
+
+        return func(content, **kwargs)
+    return wrapper
+
+
+@jwk_from_bytes_argument_conversion
+def jwk_from_private_bytes(
+    content: bytes,
+    *,
+    serializer_load_function_private:
+        Union[str, Callable[[bytes, Optional[str], object], object]] = None,
```

Comment:
`serializer_load_function_private` should not be optional.

Reply:
it isn't, but it's a keyword argument, if you call it with `None` then the decorator raises an exception
The typing is `Union[str, Callable[[bytes, Optional[str], object], object]]` and not `Optional[Union[...]]`

it can become a positional argument, but to keep consistency with jwk_from_bytes where positional arguments can be very confusing, We made this with [PEP3102](https://www.python.org/dev/peps/pep-3102/)

Reply:
I understand your intent of compatibility with `jwk_from_bytes` and it's OK to leave this keyword parameter.

Reply:
Sorry, but I've changed the mind. This should be positional and required parameter.

- the parameter is always required
- callers which import this not exported at the top function explicitly must know what the parameters are
- swapping `jwk_from_*_bytes` seems to be very rare
- `jwk_from_public_bytes(content, public key loader)` and `jwk_from_private_bytes(content, private key loader)`are consistent in number of parameters and their types.
- for readability, it's enough to use consistent name between `jwk_from_bytes` and `jwk_from_public_bytes/jwk_from_private_bytes` and call with that parameter name on the caller side.

Reply:
Or just make `serializer_load_function_private` required, by removing `= None` default value is better.